io/src/main/java/org/commonjava/maven/ext/manip/rest/DefaultVersionTranslator.java

```diff
@@ -19,116 +19,172 @@
-import java.util.Map;
-import java.util.Random;
+import java.util.*;
 
 import static org.apache.commons.lang.StringUtils.isNotEmpty;
 
 /**
  * @author vdedik@redhat.com
+ * @author Jakub Senko <jsenko@redhat.com>
  */
 public class DefaultVersionTranslator
-    implements VersionTranslator
-{
+        implements VersionTranslator {
     private static final Random RANDOM = new Random();
 
-    private final Base32 codec = new Base32( );
+    private static final Base32 CODEC = new Base32();
 
-    private final Logger logger = LoggerFactory.getLogger( getClass() );
+    private final Logger logger = LoggerFactory.getLogger(getClass());
 
     private final String endpointUrl;
+    private final int chunkSplitCount;
 
-    /**
-     * If 0 - no limit.
-     * Otherwise loop in batches of maxRestSize.
-     */
-    private final int maxRestSize;
+    public DefaultVersionTranslator(String endpointUrl) {
+        this(endpointUrl, 4);
+    }
 
-    public DefaultVersionTranslator( String endpointUrl, int maxRestSize )
-    {
+    public DefaultVersionTranslator(String endpointUrl, int chunkSplitCount) {
         this.endpointUrl = endpointUrl;
-        this.maxRestSize = maxRestSize;
-
-        Unirest.setObjectMapper( new ProjectVersionRefMapper() );
+        this.chunkSplitCount = chunkSplitCount;
+        Unirest.setObjectMapper(new ProjectVersionRefMapper());
         // According to https://github.com/Mashape/unirest-java the default connection timeout is 10000
         // and the default socketTimeout is 60000.
         // We have increased the first to 30 seconds and the second to 10 minutes.
-        Unirest.setTimeouts( 30000, 600000 );
+        Unirest.setTimeouts(30000, 600000);
     }
 
-    public Map<ProjectVersionRef, String> translateVersions( List<ProjectVersionRef> allProjects )
-    {
-        final List<List<ProjectVersionRef>> partition = ListUtils.partition( allProjects, (maxRestSize == 0 ? allProjects.size() : maxRestSize) );
+
+    /**
+     * Translate the versions.
+     * There may be a lot of them, possibly causing timeouts or other issues.
+     * This is mitigated by splitting them into smaller chunks when an error occurs and retrying.
+     */
+    public Map<ProjectVersionRef, String> translateVersions(List<ProjectVersionRef> projects) {
         final Map<ProjectVersionRef, String> result = new HashMap<>();
+        logger.debug("Translating versions: " + projects);
+        final Queue<Task> queue = new ArrayDeque<>();
+        queue.add(new Task(projects, chunkSplitCount, endpointUrl));
+
+        while (!queue.isEmpty()) {
+            Task task = queue.remove();
+            task.executeTranslate();
+            if (task.isSuccess()) {
+                result.putAll(task.getResult());
+            } else {
+                if (task.canSplit()) {
+                    logger.debug("Failed to translate versions, splitting and retrying. " +
+                            "Chunk size is " + task.getChunkSize());
```

Comment:
Is this going to output the previous chunk size rather than what it split it into?